icu4j/main/classes/core/src/com/ibm/icu/message2/Mf2DataModelFormatter.java

```diff
@@ -0,0 +1,277 @@
+        if (func == null) {
+            func = customFunctionRegistry.getFormatter(functionName);
+        }
+        return func;
+    }
+
+    String format(Map<String, Object> arguments) {
+        Pattern patternToRender;
+        List<Expression> selectors = dm.getSelectors();
+        if (selectors.isEmpty()) {
+            patternToRender = dm.getPattern();
+        } else {
+            // Collect all the selector functions in an array, to reuse
+            List<Selector> selectorFunctions = new ArrayList<>(selectors.size());
+            for (Expression selector : selectors) {
+                String functionName = selector.getFunctionName();
+                SelectorFactory funcFactory = standardFunctionRegistry.getSelector(functionName);
+                if (funcFactory == null) {
+                    funcFactory = customFunctionRegistry.getSelector(functionName);
+                }
+                if (funcFactory != null) {
+                    Map<String, Object> opt = new HashMap<>();
+                    for (Entry<String, Value> e : selector.getOptions().entrySet()) {
+                        opt.put(e.getKey(), variableToObjectEx(e.getValue(), arguments));
+                    }
+                    selectorFunctions.add(funcFactory.createSelector(locale, opt));
+                } else {
+                    throw new IllegalArgumentException("Unknown selector type: " + functionName);
+                }
+            }
+            // This should not be possible, we added one function for each selector, or we have thrown an exception.
+            // But just in case someone removes the throw above?
+            if (selectorFunctions.size() != selectors.size()) {
+                throw new IllegalArgumentException("Something went wrong, not enough selector functions, "
+                        + selectorFunctions.size() + " vs. " + selectors.size());
+            }
+
+            patternToRender = null;
+            // Iterate "vertically", through all variants
+            for (Entry<SelectorKeys, Pattern> variant : dm.getVariants().entrySet()) {
+                int maxCount = selectors.size();
+                List<String> keysToCheck = variant.getKey().getKeys();
+                if (selectors.size() != keysToCheck.size()) {
+                    throw new IllegalArgumentException("Mismatch between the number of selectors and the number of keys: "
+                            + selectors.size() + " vs. " + keysToCheck.size());
+                }
+                boolean matches = true;
+                // Iterate "horizontally", through all matching functions and keys
+                for (int i = 0; i < maxCount; i++) {
+                    Expression selector = selectors.get(i);
+                    String valToCheck = keysToCheck.get(i);
+                    Selector func = selectorFunctions.get(i);
+                    Map<String, Object> options = mf2OptToFixedOptions(selector.getOptions());
```

Comment:
Questions... I thought we fetched fixed options for the factory and variable options for the actual selector. ??

Reply:
Added TODO

Reply:
Done
And added unit tests.